oldtest: support for running by filename

blueyed opened the pull request:
Follow-up to 8969efca8 (vim-patch:8.1.0723)

NOTE: This changes the main entrypoint for running single oldtest files
to not use/require the ".res" extension anymore.  But it is handled for
B/C.

Adds a phony rule to run oldtest by filename.

testdir/Makefile: only keep files in NEW_TESTS_IN_ALOT

Taken out of https://github.com/neovim/neovim/pull/10650.

blueyed committed:
oldtest: support for running by filename

Follow-up to 8969efca8 (vim-patch:8.1.0723)

NOTE: This changes the main entrypoint for running single oldtest files
to not use/require the ".res" extension anymore.  But it is handled for
B/C.

Adds a phony rule to run oldtest by filename.

testdir/Makefile: only keep files in NEW_TESTS_IN_ALOT

justinmk commented on test/README.md:
```diff
@@ -77,11 +77,10 @@ To run all legacy Vim tests:
+To run a *single* legacy test file:
 
-    TEST_FILE=test_syntax.res make oldtest
+    make src/nvim/testdir/test_syntax.vim
```
is TEST_FILE not supported now? It makes more sense  to recommend that if it still works, otherwise people have to remember different ways of running tests for Lua vs oldtest.

blueyed replied:
ok, will change that to TEST_FILE=test_syntax.vim then.
Maybe we can mention both though, if you would like to have support for the full name.

blueyed replied:
@justinmk 
Should I mention both here?

justinmk replied:
> Should I mention both here?

If the plan is to make it work for Lua tests also, then I guess that makes sense.  In general though too many options makes docs hard to read.

blueyed replied:
ok, amended.

justinmk commented on Makefile:
```diff
@@ -119,8 +119,14 @@ oldtest: | nvim build/runtime/doc/tags
+# Build oldtest by specifying the .vim filename.
+.PHONY: phony_force
+src/nvim/testdir/%.vim: phony_force
+	$(MAKE) oldtest SCRIPTS= NEW_TESTS_RES=$(patsubst src/nvim/testdir/%.vim,%.res,$@)
```
Specifying the full `src/nvim/testdir/%.vim` path is not worth all this complication. If there's a way we can support this:

    TEST_FILE=foo.vim make oldtest

then that's good enough, and strongly preferred if it avoids a bunch of extra Makefile tricks.

blueyed replied:
I think we can support both, and thought that the full path is better, since it allows for tab completion of filenames, and also goes into the direction of the `test/run.sh` script I suggested somewhere, and where you've said it would be better to have it in / via the Makefile.

justinmk replied:
> also goes into the direction of the `test/run.sh` script I suggested somewhere, and where you've said it would be better to have it in / via the Makefile.

`test/run.sh` is yet more indirection... definitely better to keep things in the Makefile, if it's something that's really needed at all. But that's also a compromise because CMake is the build system, not make...

I also am uncomfortable with gold-plating "oldtest" in general. `TEST_FILE=foo.vim` is good to have for parallel form with the Lua tests. That's the only reason.

blueyed replied:
> But that's also a compromise because CMake is the build system, not make...

Not for oldtests though.  Here our main Makefile redirects to the local Makefile there.

> I also am uncomfortable with gold-plating "oldtest" in general. `TEST_FILE=foo.vim` is good to have for parallel form with the Lua tests. That's the only reason.

It's only done there first.  The same pattern would also work then for other tests in general.

justinmk replied:
> Not for oldtests though. Here our main Makefile redirects to the local Makefile there.

That's only because no one converted it to a proper CMake target. (Though we did so with many other Vim Makefiles, including docs IIRC.)

blueyed commented on Makefile:
```diff
@@ -201,10 +207,10 @@ lint: check-single-includes clint lualint _opt_pylint _opt_shlint
 
-$(DEPS_BUILD_DIR)/%:
+$(DEPS_BUILD_DIR)/%: phony_force
 	$(BUILD_CMD) -C $(DEPS_BUILD_DIR) $(patsubst $(DEPS_BUILD_DIR)/%,%,$@)
```
Note that the "phony_force" below are not really necessary here, but make sense for any "%" rule, so I think it's ok to have them added via this also.

blueyed commented on src/nvim/testdir/Makefile:
```diff
@@ -114,7 +115,7 @@ fixff:
 
 # Execute an individual new style test, e.g.:
-# make test_largefile
+# 	make test_largefile
```
This is to align with Vim's file.

blueyed committed:
src/nvim/testdir/%.vim: use same rule as above

Not going through "$(MAKE)" avoids GNUmakefile being used then (which I
use for WIP things), and it seems like SINGLE_MAKE should be used anyway
probably.

blueyed committed:
reformat

blueyed committed:
NEW_TESTS: all test files

blueyed committed:
fixup! reformat
